---
title: "time-scaled haplotype density"
output: html_document
date: '2022-06-19'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(thd)
library(ape)
library(tidyverse)
library(ggsci)

msa_dir = '/labs/jandr/walter/tb/pgy/msa/'
xml_dir =  '/labs/jandr/walter/tb/pgy/xml/'
plot_dir = '/labs/jandr/walter/tb/pgy/plots/'
meta_dir = '/labs/jandr/walter/tb/pgy/metadata/'

save.image('/labs/jandr/walter/tb/pgy/thd.RData')
```
```{r read in files}
# Metadata
meta_file = paste(meta_dir, 'pgy_full_data.csv', sep = '')
pgy_df <- read_csv(meta_file)

# SNP MSA
snps_fasta <- '/labs/jandr/walter/tb/pgy/msa/pgy_msa_snps.fa'
# Need to use read.dna function so that DNAbin has row names
f <- read.dna(snps_fasta,format = 'fasta') # To get the characters, use, as.character = TRUE, to input matrix, use, as.matrix = TRUE. (Format will not be DNAbin)
str(f)

# Strange pattern where '1' is added to end of sample name.
new_labels <- str_remove(labels(f), "1$")
rownames(f) <- new_labels
rownames(f)[which(!rownames(f) %in% pgy_df$seq_name)] # all fasta names are in seq_name


```

```{r thd, following the tutorial}

# Get pairwise distances
d <- dist.dna(f, pairwise.deletion = TRUE, model = 'N', as.matrix = TRUE)
hist(d)

tshort <- 20
tlong <- 200
genome_length <- 4411532
masked_length <- 495091
m <- genome_length - masked_length # effective genome sizedim(f)[2] # unsure what this should be
m
mu <- 1e-7
tmp = as.matrix(d)[1:400,1:400]
# Compute THD values with short and long timescales
thd.short <- thd(d, tshort, m, mu)
thd.long  <- thd(d, tlong, m, mu)
par(mfrow = c(1, 2))
hist(thd.short, xlab = "Short-term THD", main = "")
hist(thd.long, xlab = "Long-term THD", main = "")

# Scale with z-scores
thd.short.lz <- scale(log(thd.short))
thd.long.lz <- scale(log(thd.long))

# Combine thd with metadata.
thd_df <- data.frame(cbind(thd.short.lz,thd.long.lz)) %>%
  rownames_to_column(var = 'seq_name') %>%
  rename(thd.short = X1, thd.long = X2) %>% 
  left_join(pgy_df %>% filter(include == 1),
            by = 'seq_name') 

table(thd_df$prisoner, useNA = 'always')
dim(thd_df)

# Examine the distribution of THDs depending on the timescale
thd.plot <- thd_df %>%
  pivot_longer(cols = c(thd.short,thd.long),names_to = 'Time period', values_to = 'thd') %>%
  mutate(`Time period` = case_when(`Time period` == 'thd.long' ~ 'Long',
                                   `Time period` == 'thd.short' ~ 'Short'),
         `Time period` = factor(`Time period`, levels = c('Short','Long'))) %>%
  ggplot(aes(x = inc_status, y = thd, fill = inc_status)) + 
  geom_boxplot() + facet_wrap(~`Time period`) +
  theme_classic() + ylim(-2,1) + ylab('Time-scaled haplotype density') + 
  xlab('Incarceration status at TB notification') +
  scale_fill_lancet(name="Incarceration status") + 
  theme(axis.text.x=element_blank())
thd.plot
#ggsave(thd.plot, filename = paste(plot_dir, 'thd_plot.pdf', sep = ''))

# Summarize statistics for reporting
thd_df %>%
  group_by(inc_status) %>%
  dplyr::summarize(median(thd.short),
            quantile(thd.short, .25), 
            quantile(thd.short, .75)) %>%
  mutate(across(where(is.numeric), round, 2))

# t-test compare incarcerated group to the formerly incarcerated
thd_df %>%
  filter(inc_status %in% c('Incarcerated','Formerly incarcerated')) %>%
  t_test(formula = thd.short ~ inc_status,
      order = c('Incarcerated','Formerly incarcerated'),
      alternative = "greater") %>%
  mutate(across(where(is.numeric), round, 2))

# t-test compare incarcerated group to the community
thd_df %>%
  filter(inc_status %in% c('Incarcerated','Community')) %>%
  t_test(formula = thd.short ~ inc_status,
      order = c('Incarcerated','Community'),
      alternative = "greater") %>%
  mutate(across(where(is.numeric), round, 5))

# Summarize statistics for isn resistance
thd_df %>%
  group_by(isoniazid) %>%
  dplyr::summarize(n = n(), 
                   median(thd.short),
            quantile(thd.short, .25), 
            quantile(thd.short, .75)) %>%
  mutate(across(where(is.numeric), round, 2))

# t-test compare ahpC isn resistance to none
thd_df %>%
  filter(isoniazid %in% c('ahpC_c.-74G>A','-')) %>%
  t_test(formula = thd.short ~ isoniazid,
      order = c('-','ahpC_c.-74G>A'),
      alternative = "greater") %>%
  mutate(across(where(is.numeric), round, 5))

```
```{r adjusted THD}
adj.short <- thd.adjust(thd.short.lz, d, 20)
adj.long <- thd.adjust(thd.long.lz, d, 20)
ncol(adj.short); ncol(adj.long)
dim(adj.short)

# Combine adjusted THD with dataframe
thd_adj_df <- data.frame(cbind(adj.short,adj.long)) %>%
  rownames_to_column(var = 'seq_name') %>%
  rename(adj.short = X1, adj.long = X2) %>%
  left_join(thd_df, by = 'seq_name')

# Anova to test if effect of incarceration is independent of population structure - short timescale
anova(lm(thd.short ~ inc_status, data = thd_df))

# Anova to test if effect of incarceration is independent of population structure - short timescale
anova(lm(thd.short ~ inc_status + adj.short, data = thd_df))

# Anova to test if effect of incarceration is independent of population structure - long timescale
anova(lm(thd.long ~ inc_status + location, data = thd_df))

# Anova to test if effect of incarceration is independent of population structure - long timescale
anova(lm(thd.long ~ inc_status + adj.long + location, data = thd_df))

# Anova to test if effect of isn resistance is independent of population structure - short timescale
anova(lm(thd.short ~ isoniazid, data = thd_df))

# Anova to test if effect of isn resistance is independent of population structure - short timescale - association
anova(lm(thd.short ~ inc_status + isoniazid + adj.short, data = thd_df))

# But not drug-resistance overall
anova(lm(thd.short ~ DR_type + adj.short, data = thd_df))


```